pandas/core/arrays/arrow/array.py

```diff
@@ -749,6 +755,32 @@ def take(
                 indices_array[indices_array < 0] += len(self._data)
             return type(self)(self._data.take(indices_array))
 
+    @doc(ExtensionArray.to_numpy)
+    def to_numpy(
+        self,
+        dtype: npt.DTypeLike | None = None,
+        copy: bool = False,
+        na_value: object = lib.no_default,
+    ) -> np.ndarray:
+        # TODO: copy argument is ignored
+
+        if na_value is lib.no_default:
+            na_value = self.dtype.na_value
+
+        pa_type = self._data.type
+        if pa.types.is_timestamp(pa_type) or pa.types.is_duration(pa_type):
+            result = np.array(self.tolist(), dtype=np.object_)
```

Comment:
Can we generally not use `self._data.to_numpy()`?: https://arrow.apache.org/docs/python/generated/pyarrow.Array.html#pyarrow.Array.to_numpy

Reply:
I've looked at it and I think it would only be usable for a subset of cases. The inability to specify `dtype` and `na_value` make it difficult for general usage in this interface. Also, need to be careful with things like timestamps as it will strip off the timezone.

Reply:
Gotcha. Yeah where `EA.to_numpy()` has more functionality (dtype and na_value) than pyarrow's `to_numpy()` I agree we need to have out custom behavior.

But otherwise I would lean towards using `self._data.to_numpy()`. IMO at this stage of the ArrowExtensionArray it's better to let pyarrow behavior pass through (and potentially improve) rather than modifying it's behavior and possibly diverging

Reply:
Agreed with trying to limit divergence from pyarrow.

I was attempting to maintain status quo behavior with better perf here. Current default behavior of `ArrowExtensionArray.to_numpy` is mostly consistent with other masked/nullable pandas dtypes. If its preferred to let the pyarrow behavior flow through the expected behavior in a number of tests would need to be changed (or xfail)

Here is an example where behavior differs between the current `ArrowExtensionArray.to_numpy` and pyarrow's to_numpy:

```
import pandas as pd

data = [1, None]

pd.array(data, dtype="Int64").to_numpy()                    # -> array([1, <NA>], dtype=object)
pd.array(data, dtype="int64[pyarrow]").to_numpy()           # -> array([1, <NA>], dtype=object)
pd.array(data, dtype="int64[pyarrow]")._data.to_numpy()     # -> array([1.0, nan])
```

Let me know if you have thoughts or suggestions. Happy to modify or close this if you think it needs more consideration.

Reply:
I'll also note that `np.array(pa_array)` goes through pyarrow's to_numpy method. pyarrow implements `pa.Array.__array__` which is a small wrapper around the to_numpy method. 

So these should be equivalent and both go through `pa.Array.to_numpy`:
```
np.array(pa_array)
pa_array.to_numpy(zero_copy_only=False, writable=True)
```

Reply:
Okay in the spirit of maintaining the status quo, discussion on whether to rely on `pa.Array.to_numpy()` exclusively can happen later.

But with respect to consistency with the `BaseMaskedArray.to_numpy()`, `ArrowExtensionArray` never inherited from `BaseMaskedArray`, so the status quo would have been to match `EA.to_numpy` and not coerce to object like `BaseMaskedArray` I think as this would be an API change for `ArrowExtensionArray`

Reply:
Apologies if I've misinterpreted your comments. The cast to `object` is existing behavior on main and looks like it may have always been the case.

By default (on main), `ArrowStringArray.to_numpy` (inherited from ExtensionArray) results in a number of object casts which `ArrowStringArray._data.to_numpy` would otherwise not cast to `object`. I wasn't suggesting `ArrowStringArray` should behave like `BaseMaskedArray`, but just observing that it often does in its current state. 

Here are a few examples on the *main* branch:

```
import pandas as pd

arr = pd.array([1, None], dtype="int64[pyarrow]")
arr.to_numpy()        # -> array([1, <NA>], dtype=object)
arr._data.to_numpy()  # -> array([ 1., nan])


arr = pd.array([pd.Timestamp("2000-01-01")], dtype='timestamp[ns][pyarrow]') 
arr.to_numpy()        # -> array([Timestamp('2000-01-01 00:00:00')], dtype=object)
arr._data.to_numpy()  # -> array(['2000-01-01T00:00:00.000000000'], dtype='datetime64[ns]')


arr = pd.array([pd.Timestamp("2000-01-01"), None], dtype='timestamp[ns][pyarrow]') 
arr.to_numpy()        # -> array([Timestamp('2000-01-01 00:00:00'), <NA>], dtype=object)
arr._data.to_numpy()  # -> array(['2000-01-01T00:00:00.000000000', 'NaT'], dtype='datetime64[ns]')


arr = pd.array([1], dtype='duration[ns][pyarrow]') 
arr.to_numpy()        # -> array([Timedelta('0 days 00:00:00.000000001')], dtype=object)
arr._data.to_numpy()  # -> array([1], dtype='timedelta64[ns]')


arr = pd.array([1, None], dtype='duration[ns][pyarrow]') 
arr.to_numpy()        # -> array([Timedelta('0 days 00:00:00.000000001'), <NA>], dtype=object)
arr._data.to_numpy()  # -> array([    1, 'NaT'], dtype='timedelta64[ns]')
```

So I guess one question is what the defualt behavior of `ArrowStringArray.to_numpy` should be when a user does not pass dtype or na_value? At the moment, all nulls are filled with `dtype.na_value` which is `pd.NA` for all pyarrow backed dtypes.

Reply:
Something like this would allow the pyarrow to_numpy behavior to flow through by default for all but a few cases. This passes all tests, but it does result in different behavior that I guess isn't tested. It also lets us implement the `copy` parameter. 

```
    def to_numpy(
        self,
        dtype: npt.DTypeLike | None = None,
        copy: bool = False,
        na_value: object = lib.no_default,
    ) -> np.ndarray:
        pa_type = self._data.type
        if (
            is_object_dtype(dtype) 
            or pa.types.is_timestamp(pa_type) 
            or pa.types.is_duration(pa_type)
        ):
            result = np.array(list(self), dtype=dtype)
        else:
            result = np.asarray(self._data, dtype=dtype)
            if copy or na_value is not lib.no_default:
                result = result.copy()
        if na_value is not lib.no_default:
            result[self.isna()] = na_value
        return result
```

Reply:
The above implementation looks to be more permissive and better IMO. I do think naive timestamp and duration types should eventually result in `datetime64` and `timedelta64` types but that can be a follow up change

```
In [15]: pd.Series([pd.Timestamp(1)]).to_numpy()
Out[15]: array(['1970-01-01T00:00:00.000000001'], dtype='datetime64[ns]')

In [16]: pd.Series([pd.Timedelta(1)]).to_numpy()
Out[16]: array([1], dtype='timedelta64[ns]')

In [17]: pd.Series([pd.Timestamp(1)], dtype="timestamp[ns][pyarrow]").to_numpy()
Out[17]: array([Timestamp('1970-01-01 00:00:00.000000001')], dtype=object)

In [18]: pd.Series([pd.Timedelta(1)], dtype="duration[ns][pyarrow]").to_numpy()
Out[18]: array([Timedelta('0 days 00:00:00.000000001')], dtype=object)
```

Reply:
Sounds good. If we go with that change and get closer to pyarrow behavior, is it ok that (by default) `to_numpy` will no longer represent missing values with pd.NA (and cast to object)? Thinking specifically about `string[pyarrow]` which would behave differently than `string[python]` if this change were to be made.

```
In [2]: pd.array(["foo", pd.NA], dtype="string[python]").to_numpy()        
Out[2]: array(['foo', <NA>], dtype=object)

In [3]: pd.array(["foo", pd.NA], dtype="string[pyarrow]").to_numpy()        
Out[3]: array(['foo', <NA>], dtype=object)

In [4]: pd.array(["foo", pd.NA], dtype="string[pyarrow]")._data.to_numpy()
Out[4]: array(['foo', None], dtype=object)
```

Reply:
I just pushed the changes discussed above but left ArrowStringArray behavior as is with a TODO note. (see my comment directly above regarding divergence from `string[python]`).

Agreed it would be nice if duration and timestamp followed the same default behavior. When I tried it, there were a number of failing tests due to loss of timezone, non-ns units, etc. I'd suggest we leave that for a follow-up. 

Avoiding the cast to object further improves the timings in the OP.

Reply:
Agreed with tackling the duration and timestamp stuff in a followup

Reply:
Got it. updated to use pd.NA by default for missing values.